### PR TITLE
Pipes - Nkiru Onwuneme - Grocery_Store

### PR DESCRIPTION
**Pull request by nkiruka**
# Grocery Store
Congratulations! You're submitting your assignment.

## Comprehension Questions

|  Question 	|   Response	|
|---	|---	|
|   Why is it useful to put classes inside modules? 	|   Modules are useful to prevent name clashes.  Modules serves as name spacing.	|
|   What is accomplished with `raise ArgumentError`? 	| An exception is raised "when arguments passed to a method are incorrect, unexpected, or invalid in some way" (source: https://airbrake.io/blog/ruby-exception-handling/argumenterror)	|
|   Why do you think we made the `.all` & `.find` methods class methods?  Why not instance methods?	|  It prevents having to create an instance of the class each time.  The class methods are not tied to any instance of the class.	|
|   Why does it make sense to use inheritance for the online order?  | There is a "is_a" relationship between order and online order, because online_order is a type of order that can be placed, other examples are store orders, magazine order etc.  So since online_order is a subclass, it makes sense that it will inherit some properties of order, therefore, eliminating code duplication. |
|   Did the presence of automated tests change the way you thought about the problem? How? |Yes, it forced me to think about the design of the program, instead of writing unnecessary code that might have be deleted later.  TDD also made me think of edge cases that I would have liked to test to further optimize my code. |

### CODE REVIEW

File: lib/order.rb [hunk 1]
```diff
@@ -8,11 +11,46 @@ def initialize(id, products)
...
+
+    def self.all
+      orders = []
+      CSV.open("./support/orders.csv", "r").each do |line|
+        products = {}
+        id = line[0].to_i
+
+        line[1].split(";").each do |product_info|
+          data = product_info.split(":")
+          products[data[0]] = data[1].to_f
```
Comment (droberts-sea) [inline lib/order.rb:38]: I like this logic for parsing the product list - good work! If I might make a small suggestion for readability, you might do something like this:

```ruby
line[1].split(";").each do |product_info|
  name, price = product_info.split(":")
  products[name] = price.to_f
end
```

File: lib/order.rb [hunk 2]
```diff
@@ -8,11 +11,46 @@ def initialize(id, products)
...
+      end
+      return orders
+    end
+
+    def self.find(id)
+      found_order = nil
+      orders = self.all
+      orders.each do |order|
+        if order.id == id
+          return found_order = order
```
Comment (droberts-sea) [inline lib/order.rb:50]: If you immediately `return` here, do you really need the `found_order` variable?

File: lib/online_order.rb [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+    def self.find(id)
+      super(id)
+    end
+
+    def self.find_by_customer(customer_id)
+      customer_orders = []
+      online_orders = self.all
+      if online_orders.any? {|instance| instance.customer.customer_id == customer_id_find.to_s}
+        customers.each do |order|
```
Comment (droberts-sea) [inline lib/online_order.rb:61]: We talked about this offline, but I'll recap here:

This is getting close, but I suspect you're overcomplicating things. Here we're looking for all the `OnlineOrder`s where the customer ID matches. It'll look a lot like `OnlineOrder.find`, but it will select several instances instead of just one.

File: lib/online_order.rb [hunk 4]
```diff
@@ -0,0 +1,82 @@
+require 'csv'
+require_relative 'order'
+
+
+module Grocery
+  class OnlineOrder < Order
+    attr_reader :status, :customer_id, :customer
+
+    def initialize(id, products, customer_id, status = :pending)
+      super(id, products)
```
Comment (droberts-sea) [inline lib/online_order.rb:10]: You've done a good job of using inheritance and `super` in this class to keep your code DRY.

File: specs/online_order_spec.rb [hunk 5]
```diff
@@ -10,62 +14,159 @@
...
+      expected_total = (sum + (sum * 0.075).round(2))
+
+      online_order.total.must_equal expected_total
     end
   end
 
   describe "#add_product" do
+    it "Does not permit action for processing status" do
+      status = :completed
+      customer_id = 23
```
Comment (droberts-sea) [inline specs/online_order_spec.rb:83]: The name of this test talks about the "processing status", but then you set `status = :completed`. That seems like a mismatch to me. Copy-paste error?

File: specs/order_spec.rb [hunk 6]
```diff
@@ -78,32 +78,84 @@
...
-      # Feel free to split this into multiple tests if needed
+      orders = Grocery::Order.all
+      orders.must_be_kind_of Array
+    end
+
+    it "checks if orders match what is the CSV file" do
+      orders = Grocery::Order.all
+      csv = CSV.read("./support/orders.csv")
+      csv_product = csv.to_a[18][1].gsub(/:|;/, " ")
+      order_product = orders[18].products.flatten.join(" ")
```
Comment (droberts-sea) [inline specs/order_spec.rb:93]: This is a very concise way to make this check!

As we talked about offline, this bit of logic is pretty complex. It would probably be worthwhile to leave some comments here so that you (or your coworkers) have an idea of what format the data is in, and what your transformations are doing.

**Comment by droberts-sea** [comment]
# Grocery Store
## What We're Looking For

| Feature | Feedback|
|---	|---	|
| **Baseline** | |
| Answered comprehension questions	| yes |
| Used Git Regularly	| Your commit cadence is good! However, I would like to see more descriptive commit messages. "Edited test #5" doesn't tell me much about what changed with that commit or why. Something like "Added basic implementation of Order.find" or "Fixed a bug where customer ID was incorrectly handled" would be much better. The goal is that someone should be able to look at your commit history and get a good idea of how work on your project progressed without having to read your code at all. |
| **Wave 1** | |
| All provided tests pass | yes |
| Using the appropriate `attr_` for instance variables | yes |
| **Wave 2** | |
| All stubbed tests are implemented fully and pass | yes |
| Appropriately parses the product data from CSV file in `Order.all` | yes |
| Used `CSV` library only in `Order.all` (not in `Order.find`) | yes |
| Used `Order.all` to get account list in `Order.find`	| yes |
| **Wave 3** | |
| All stubbed tests are implemented fully and pass | missing tests for `OnlineOrder.find_by_customer` |
| Used inheritance in the initialize for online order | yes |
| Used inheritance for the total method in online order | yes |
| Appropriately created Customer class and class methods `all` & `find` | yes |
| **Additional Notes**	| |

Great job overall! I've identified a few places below where you could make things a little clearer, but in general this is a solid submission. Keep up the good work!
